### DISCUSSION TITLE
Cross-Compiling AWS IoT Device SDK C++ for Armhf

### DISCUSSION THREAD

**Discussion by zlberto** [aws/aws-iot-device-sdk-cpp-v2#683] [General]
Hello everyone,

I'm currently working on cross-compiling the AWS IoT Device SDK C++ for a Raspberry Pi Zero 2W project. I've encountered some challenges, particularly around using a custom-built OpenSSL library for the ARM architecture.

## Environment:
Target Device: Raspberry Pi Zero 2W
Host System: MacOS M2 running Ubuntu 22.04 docker container
Cross-Compilation Toolchain: arm-linux-gnueabihf
AWS IoT Device SDK C++ Version: (SDK version or commit hash)
CMake Version: 3.22.1

### Issue Description:
The main issue arises during the linking phase, where the SDK or one of its dependencies (specifically s2n-tls) is unable to find or correctly link against the custom-built OpenSSL (libcrypto.so.1.0.0 not found).

Despite setting the -DUSE_OPENSSL=ON flag and providing paths to the custom-built OpenSSL (using -DOPENSSL_ROOT_DIR, -DOPENSSL_CRYPTO_LIBRARY, and -DOPENSSL_SSL_LIBRARY), the build process seems to be not recognizing or correctly using the provided OpenSSL library paths.

### Steps Taken:
Cross-compiled OpenSSL for ARM and installed it in a custom directory.

Attempted to cross-compile AWS IoT Device SDK C++ with the following CMake configuration:

```bash
cmake -DBUILD_SHARED_LIBS=ON \
  -DCMAKE_SYSTEM_NAME=Linux \
  -DCMAKE_SYSTEM_PROCESSOR=arm \
  -DCMAKE_CXX_COMPILER=${CROSS_COMPILE}-g++ \
  -DCMAKE_C_COMPILER=${CROSS_COMPILE}-gcc \
  -DCMAKE_LINKER=${CROSS_COMPILE}-ld \
  -DCMAKE_AR=${CROSS_COMPILE}-ar \
  -DCMAKE_INSTALL_PREFIX=/sdk/aws-iot-device-sdk-cpp-v2-build \
  -DBUILD_DEPS=ON \
  -DCMAKE_BUILD_TYPE="Debug" \
  -DUSE_OPENSSL=ON \
  -DOPENSSL_ROOT_DIR=/path/to/local/openssl \
  -DOPENSSL_CRYPTO_LIBRARY=/path/to/local/openssl/lib/libcrypto.so \
  -DOPENSSL_SSL_LIBRARY=/path/to/local/openssl/lib/libssl.so \
  /sdk/aws-iot-device-sdk-cpp-v2
```

```
CMake Error at /usr/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find crypto (missing: crypto_LIBRARY)
Call Stack (most recent call first):
  /usr/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  crt/aws-crt-cpp/crt/s2n/cmake/modules/Findcrypto.cmake:75 (find_package_handle_standard_args)
  crt/aws-crt-cpp/crt/s2n/CMakeLists.txt:233 (find_package)
```

**Comment by bretambrose** [comment]
In general we often cannot offer much assistance with cross-compile issues unless there's definitive proof that it's a Cmake config error on our part.

Two things do jump out at me though:

1. I would advise using CMAKE_PREFIX_PATH rather than setting openssl paths directly: `-DCMAKE_PREFIX_PATH=/path/to/local/openssl`
2. Make sure you're building against a raspberry Pi sysroot.  Maybe your toolchain is configured to use a Pi sysroot, but I dont see anything hinting at it in the description.  Either use a sysroot corresponding to your target (Pi) or use a Pi container that matches your target baseline.
